power_grid_model_c/power_grid_model/include/power_grid_model/auxiliary/meta_data.hpp

```diff
@@ -14,247 +14,279 @@
-        *reinterpret_cast<RealValue<false>*>(ptr) = RealValue<false>{nan, nan, nan};
-    };
-    static constexpr CheckNaNFunc check_nan = [](RawDataConstPtr ptr) -> bool {
-        return is_nan(*reinterpret_cast<RealValue<false> const*>(ptr));
-    };
-    static constexpr SetValueFunc set_value = set_value_template<RealValue<false>>;
-    static constexpr CompareValueFunc compare_value = [](RawDataConstPtr ptr_x, RawDataConstPtr ptr_y, double atol,
-                                                         double rtol) -> bool {
-        RealValue<false> const x = *reinterpret_cast<RealValue<false> const*>(ptr_x);
-        RealValue<false> const y = *reinterpret_cast<RealValue<false> const*>(ptr_y);
-        return (abs(y - x) < (abs(x) * rtol + atol)).all();
-    };
+struct ctype_t<RealValue<false>, false> {
+    static constexpr const char* value = "double[3]";
 };
-
 template <class T>
-struct data_type<T, true> : data_type<std::underlying_type_t<T>> {};
+struct ctype_t<T, true> : ctype_t<std::underlying_type_t<T>> {};
+template <class T>
+constexpr const char* ctype_v = ctype_t<T>::value;
+
+// set nan
+inline void set_nan(double& x) {
+    x = nan;
+}
+inline void set_nan(IntS& x) {
+    x = na_IntS;
+}
+inline void set_nan(ID& x) {
+    x = na_IntID;
+}
+inline void set_nan(RealValue<false>& x) {
+    x = RealValue<false>{nan};
+}
+template <class Enum>
+requires std::same_as<std::underlying_type_t<Enum>, IntS>
+inline void set_nan(Enum& x) {
+    x = static_cast<Enum>(na_IntS);
+}
+
+using RawDataPtr = void*;             // raw mutable data ptr
+using RawDataConstPtr = void const*;  // raw read-only data ptr
 
-struct DataAttribute {
+// meta attribute
+template <class StructType, auto member_ptr>
+struct MetaAttributeImpl {
+    using ValueType = typename trait_pointer_to_member<decltype(member_ptr)>::value_type;
+    static bool check_nan(RawDataConstPtr buffer_ptr, Idx pos) {
+        return is_nan((reinterpret_cast<StructType const*>(buffer_ptr) + pos)->*member_ptr);
+    }
+    static void set_value(RawDataPtr buffer_ptr, RawDataConstPtr value_ptr, Idx pos) {
+        (reinterpret_cast<StructType*>(buffer_ptr) + pos)->*member_ptr = *reinterpret_cast<ValueType const*>(value_ptr);
+    }
+    static void get_value(RawDataConstPtr buffer_ptr, RawDataPtr value_ptr, Idx pos) {
+        *reinterpret_cast<ValueType*>(value_ptr) = (reinterpret_cast<StructType const*>(buffer_ptr) + pos)->*member_ptr;
+    }
+    static bool compare_value(RawDataConstPtr ptr_x, RawDataConstPtr ptr_y, double atol, double rtol, Idx pos) {
+        ValueType const& x = (reinterpret_cast<StructType const*>(ptr_x) + pos)->*member_ptr;
+        ValueType const& y = (reinterpret_cast<StructType const*>(ptr_y) + pos)->*member_ptr;
+        if constexpr (std::is_same_v<ValueType, double>) {
+            return std::abs(y - x) < (std::abs(x) * rtol + atol);
+        }
+        else if constexpr (std::is_same_v<ValueType, RealValue<false>>) {
```

Comment:
Changed. But what's the benefit of using `std::same_as` here? I can understand its advantage in the `requires` block to avoid using `enable_if`. But what is the benefit to use in `if constexpr`?